presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,374 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.elasticsearch;
+
+import com.facebook.presto.spi.ColumnMetadata;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.spi.type.RowType;
+import com.facebook.presto.spi.type.Type;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Closer;
+import io.airlift.json.ObjectMapperProvider;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+import org.elasticsearch.action.admin.cluster.shards.ClusterSearchShardsRequest;
+import org.elasticsearch.action.admin.cluster.shards.ClusterSearchShardsResponse;
+import org.elasticsearch.action.admin.indices.get.GetIndexRequest;
+import org.elasticsearch.action.admin.indices.mapping.get.GetMappingsRequest;
+import org.elasticsearch.client.transport.TransportClient;
+import org.elasticsearch.cluster.metadata.MappingMetaData;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.transport.client.PreBuiltTransportClient;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+
+import static com.facebook.presto.elasticsearch.ElasticsearchErrorCode.ELASTICSEARCH_CORRUPTED_MAPPING_MEDATA;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Map.Entry;
+import static java.util.Objects.requireNonNull;
+
+public class ElasticsearchClient
+{
+    private static final Logger LOG = Logger.get(ElasticsearchClient.class);
+
+    private final ObjectMapper objecMapper = new ObjectMapperProvider().get();
+    private final Map<SchemaTableName, ElasticsearchTableDescription> tableDescriptions;
+    private final Map<String, TransportClient> clients = new HashMap<>();
+    private final Map<ElasticsearchTableDescription, List<ColumnMetadata>> tableColumns = new HashMap<>();
```

Comment:
shouldn't be this some kind of cache? It can only grow, it would be nice if it could also shrink.

Reply:
I have no strong opinion. I would prefer cache with a ETL configuration property (with default like 30m). I am not sure how others would use this connector, maybe some people might have a lot of columns and that might become a problem. Or maybe columns information (indices) could change and it would be good to reload them.

Reply:
To do caching, I think we have to load per table/index. All columns belongs to a table will either in cache, or sent another mapping request to ELK. From my experience, ELK index's mapping information not update quite often. Refreshing it might not be very helpful.

If one index has huge number of fields, we have to keep all these fields information in memory. Refreshing it might need additional networking/parsing overhead.

One other thing is, loading all indices under one ELK cluster would be very expensive, as ELK clusters usually have a huge number of indices. 

I am inclined to think, caching is a nice to have feature, might not be very helpful right now. What do you think?